extra_geom/motors.py

```diff
@@ -0,0 +1,214 @@
+        raise ValueError(
+            "Invalid MOTORS format, expected two comma-separated integers")
+
+    positions = []
+    for q in range(n_groups):
+        try:
+            key = f"MOTOR_Q{q+1}"
+            pos = [float(n) for n in meta[key].split(',')]
+        except KeyError:
+            raise ValueError(key + " record is not found")
+        except ValueError:
+            raise ValueError(
+                f"Invalid {key} format, expected {n_motors} "
+                "comma-separated floats")
+        if len(pos) != n_motors:
+            raise ValueError(
+                f"Wrong length of {key}, expected {n_motors} floats")
+
+        positions.append(pos)
+
+    return np.array(positions)
+
+
+class BaseMotorTracker:
+    """Detector motor tracker updates geometry according to motor positions.
+    """
+    def __init__(self, ref_geom):
+        """
+        Parameters
+        ----------
+        ref_geom: one of `extra_geom.DetectorGeometryBase` implementation
+            Geometry
+        ref_motor_positions: array or sequence
+            Reference motor positions
+        """
+        self.motor_axes = self.default_motor_axes
+        self.num_groups, _, self.num_motors = self.motor_axes.shape
+        if self.num_groups != len(self.groups):
+            raise ValueError(
+                "The len of groups does not match the len of axes")
+
+        self.motor_position_shape = (self.num_groups, self.num_motors)
+        self.motor_axes_shape = self.motor_axes.shape
+
+        self.ref_geom = ref_geom
+        self.ref_motor_positions = None
+
+    @classmethod
+    def with_reference_positions(cls, ref_geom, ref_motor_positions=None):
+        tracker = cls(ref_geom)
+        if ref_motor_positions is None:
+            if hasattr(tracker.ref_geom, "motor_positions"):
+                ref_motor_positions = tracker.ref_geom.motor_positions
+            else:
+                raise ValueError(
+                    "There is no motor positions in reference geometry")
+
+        tracker.set_reference_positions(ref_motor_positions)
+        return tracker
+
+    def set_reference_positions(self, ref_motor_positions):
+        """Set reference motor positions.
+
+        Parameters
+        ----------
+        ref_motor_positions: array or list
+            New reference motor positions. The positions are expected
+            as array or list of the number of movable groups (quadrants)
+            by the number of motors per group.
+        """
+        ref_motor_positions = np.array(ref_motor_positions, copy=True)
+        if ref_motor_positions.shape != self.motor_position_shape:
+            raise ValueError(f"Expects array{self.motor_position_shape}: "
+                             f"{self.num_groups} groups moving by "
+                             f"{self.num_motors} motor each.")
+
+        self.ref_motor_positions = np.array(ref_motor_positions)
+
+    def set_motor_axes(self, new_motor_axes):
```

Comment:
Do you see a need for this to be updatable? It looks at a glance like the motor axes will always be hardcoded in the detector-specific subclass.

(It's fine if there is a potential use case, I just like to trim back APIs where possible, and it's generally easier to add things in later than to take them away)

Reply:
Now it is immutable-ish. I get rid of `with_reference_positions`, no clushes